lib/link_state.c

```diff
@@ -0,0 +1,911 @@
+	stream_put(s, &ls_pref->adv, sizeof(struct ls_node_id));
+
+	/* Push Flags, Origin & Prefix then information if there are present */
+	stream_putw(s, ls_pref->flags);
+	stream_putc(s, ls_pref->pref.family);
+	stream_putl(s, ls_pref->pref.prefixlen);
+	len = prefix_blen(&ls_pref->pref);
+	stream_put(s, &ls_pref->pref.u.prefix, len);
+	if (CHECK_FLAG(ls_pref->flags, LS_PREF_IGP_FLAG))
+		stream_putc(s, ls_pref->igp_flag);
+	if (CHECK_FLAG(ls_pref->flags, LS_PREF_ROUTE_TAG))
+		stream_putl(s, ls_pref->route_tag);
+	if (CHECK_FLAG(ls_pref->flags, LS_PREF_EXTENDED_TAG))
+		stream_putq(s, ls_pref->extended_tag);
+	if (CHECK_FLAG(ls_pref->flags, LS_PREF_METRIC))
+		stream_putl(s, ls_pref->metric);
+	if (CHECK_FLAG(ls_pref->flags, LS_PREF_SR)) {
+		stream_putl(s, ls_pref->sr.sid);
+		stream_putc(s, ls_pref->sr.sid_flag);
+		stream_putc(s, ls_pref->sr.algo);
+	}
+
+	return 0;
+}
+
+static int ls_format_msg(struct stream *s, struct ls_message *msg)
+{
+
+	/* Prepare Link State header */
+	stream_putc(s, msg->event);
+	stream_putc(s, msg->type);
+	stream_put(s, &msg->remote_id, sizeof(struct ls_node_id));
+
+	/* Add Message Payload */
+	switch (msg->type) {
+	case LS_MSG_TYPE_NODE:
+		return ls_format_node(s, msg->data.node);
+	case LS_MSG_TYPE_ATTRIBUTES:
+		return ls_format_attributes(s, msg->data.attr);
+	case LS_MSG_TYPE_PREFIX:
+		return ls_format_prefix(s, msg->data.prefix);
+	default:
+		zlog_warn("Unsupported Payload");
+		break;
+	}
+
+	return -1;
+}
+
+int ls_send_msg(struct zclient *zclient, struct ls_message *msg,
+		struct zapi_opaque_reg_info *dst)
+{
+	struct stream *s;
+	uint16_t flags = 0;
+
+	/* Check buffer size */
+	if (STREAM_SIZE(zclient->obuf) <
+	    (ZEBRA_HEADER_SIZE + sizeof(uint32_t) + sizeof(msg)))
+		return -1;
+
+	s = zclient->obuf;
+	stream_reset(s);
+
+	zclient_create_header(s, ZEBRA_OPAQUE_MESSAGE, VRF_DEFAULT);
+
+	/* Send sub-type, flags and destination for unicast message */
+	stream_putl(s, LINK_STATE_UPDATE);
+	if (dst != NULL) {
+		SET_FLAG(flags, ZAPI_OPAQUE_FLAG_UNICAST);
+		/* Send destination client info */
+		stream_putc(s, dst->proto);
+		stream_putw(s, dst->instance);
+		stream_putl(s, dst->session_id);
+	} else
+		stream_putw(s, flags);
+
+	/* Format Link State message */
+	if (ls_format_msg(s, msg) < 0) {
+		stream_reset(s);
+		return -1;
+	}
+
+	/* Put length into the header at the start of the stream. */
+	stream_putw_at(s, 0, stream_get_endp(s));
+
+	return zclient_send_message(zclient);
+}
+
+struct ls_message *ls_vertex2msg(struct ls_message *msg,
+				 struct ls_vertex *vertex)
+{
+	/* Allocate space if needed */
+	if (msg == NULL)
+		msg = XCALLOC(MTYPE_LS_DB, sizeof(struct ls_message));
+
+	msg->type = LS_MSG_TYPE_NODE;
+	msg->data.node = vertex->node;
+	msg->remote_id.origin = NONE;
+
+	return msg;
+}
+
+struct ls_message *ls_edge2msg(struct ls_message *msg, struct ls_edge *edge)
+{
+	/* Allocate space if needed */
+	if (msg == NULL)
+		msg = XCALLOC(MTYPE_LS_DB, sizeof(struct ls_message));
```

Comment:
In the case of a preexisting `msg` the content could be uninitialized mainly because `remote_id` is not fully assigned  so I propose add two lines (the same for the rest of `2msg` functions
  
```
else
               memset(msg, 0, sizeof(*msg));

```

Reply:
Done